Fix documentation comments

Julow opened the pull request:
Fixes the syntax errors reported by Odoc 2.0.2, make sure references resolve and a few other tweaks.

Odoc 2.0.2 report no warnings now :)

Julow committed:
odoc: Fix syntax errors in comments

Julow committed:
odoc: mld pages must start with an heading

Julow committed:
odoc: '@raise' and '@since' should be on their own line

Julow committed:
odoc: Fix unqualified and ambiguous references

Julow committed:
odoc: References in synopsis must be fully-qualified

Because they need to be resolved again when shown in lists of modules.

Julow committed:
odoc: Change the canonical path of stdlib modules to Stdlib.X

Odoc requires canonical paths to contain at least two parts.

Julow committed:
odoc: References inside module types must be fully qualified

References are resolved after expansion, meaning references are resolved
in the context where they are used. The context in which module types
are defined is gone at this point.

Julow committed:
api_docgen: Set dependencies between compile rules

for the stdlib.
Post-process the same rules used to build the actual stdlib.

Julow committed:
api_docgen: Fix make rules for mld files

There's no $(DOCGEN)/build.

Co-authored-by: Damien Doligez <damien.doligez@gmail.com>

Julow committed:
api_docgen: Tweak link flags for stdlib.odoc

Don't open Stdlib when linking Stdlib itself.

Julow committed:
odoc: Remove problematic references in Misc

References to 'Array.for_all' can't resolve because the 'Array' module
is shadowed and 'for_all' isn't redefined (should be included).

Co-authored-by: Damien Doligez <damien.doligez@gmail.com>

Julow commented on api_docgen/odoc/Makefile:
```diff
@@ -190,3 +194,24 @@ install:
+#       build/libref/stdlib__Y.odoc \
+#       stdlib__X.cmti
+build/.depend: ../../stdlib/.depend | build/
+	sed \
```
I noticed `sed` was already used during the build but perhaps an ocaml script would be more appropriate wrt maintenance ?

damiendoligez replied:
I think we're trying to converge on python for this kind of script.

Julow replied:
There's currently no python script in the sources, did you mean ocaml ?

Octachron replied:
We are in fact planning to *not* use python, because its installation is problematic on WIndows. Using `sed` for now is fine. 

I am wondering it we should update ocamldep to emit the right dependency information directly at some point.

Octachron commented on api_docgen/Makefile:
```diff
@@ -16,7 +16,6 @@ ROOTDIR = ..
 -include $(ROOTDIR)/Makefile.build_config
 
 ifeq ($(DOCUMENTATION_TOOL),odoc)
-  include odoc/Makefile
```
This breaks the build when `odoc` is set as the default configuration tool with `./configure --with-odoc`.

Julow replied:
I've changed it so no `include` is used and the calls from the main `Makefile` still work.
The problem with my change is that the destination of the built documentation changes to `api_docgen/ocamldoc/build/html`. Is it a problem ?

Octachron replied:
Keeping one specialized directory by build tool seems fine to me. But then you need to add an install target to the `api_docgen/Makefile` (that redirect to the odoc or ocamldoc one) and update the manual's Makefile at `manual/src/Makefile`.

Julow replied:
Sorry for the delay. I think I managed to fix the mess I made in the manual :)

Julow committed:
api_docgen: Avoid including Makefiles

The result is now built into
'api_docgen/ocamldoc/build/html' rather than
'api_docgen/html'

Co-authored-by: Damien Doligez <damien.doligez@gmail.com>

Julow committed:
api_docgen: Don't expect ROOTDIR in doc Makefiles

These Makefiles are no longer included.

damiendoligez commented on api_docgen/Makefile:
```diff
@@ -12,20 +12,33 @@
 clean:
 	rm -rf build odoc/build ocamldoc/build
+
+.PHONY: html latex man clean
```
```suggestion
.PHONY: html latex man clean install texi pdf
```

damiendoligez commented on api_docgen/odoc/Makefile:
```diff
@@ -137,17 +143,19 @@ stdlib_INDEX=\
   $(call capitalize, $(otherlibref))
 build/libref.mld:
-	echo {0 OCaml standard library} {!modules:$(stdlib_INDEX)} > $@
+	( echo "{0 OCaml standard library}"; \
+    echo "{!modules:$(stdlib_INDEX)}" ) > $@
```
indentation is wrong
```suggestion
	( echo "{0 OCaml standard library}"; \
	  echo "{!modules:$(stdlib_INDEX)}" ) > $@
```

damiendoligez commented on utils/misc.mli:
```diff
@@ -152,12 +152,12 @@ module Stdlib : sig
-    (* Same as [Array.exists], but for a two-argument predicate. Raise
-       Invalid_argument if the two arrays are determined to have
-       different lengths. *)
+    (** Same as [Array.exists], but for a two-argument predicate. Raise
+        Invalid_argument if the two arrays are determined to have
+        different lengths. *)
```
```suggestion
    (** Same as [Array.exists2] from the standard library. *)
```

damiendoligez commented on utils/misc.mli:
```diff
@@ -152,12 +152,12 @@ module Stdlib : sig
 
     val for_alli : (int -> 'a -> bool) -> 'a array -> bool
-    (** Same as {!Array.for_all}, but the
+    (** Same as [Array.for_all], but the
```
```suggestion
    (** Same as [Array.for_all] from the standard library, but the
```

damiendoligez reviewed (commented):
Looks good modulo a few remarks, but I don't feel qualified to approve the Makefile changes.

Octachron commented on stdlib/bigarray.mli:
```diff
@@ -210,7 +210,7 @@ val kind_size_in_bytes : ('a, 'b) kind -> int
 
 type c_layout = C_layout_typ (**)
-(** See {!Bigarray.fortran_layout}.*)
+(** See {!Bigarray.type-fortran_layout}.*)
```
This syntax is not supported by ocamldoc, the ocamldoc version is `{!type:Bigarray.type-fortran_layout}`.

Julow replied:
Fixed these.

Octachron commented on stdlib/bigarray.mli:
```diff
@@ -233,10 +233,10 @@ type fortran_layout = Fortran_layout_typ (**)
 
    Each layout style is identified at the type level by the
-   phantom types {!Bigarray.c_layout} and {!Bigarray.fortran_layout}
+   phantom types {!Bigarray.type-c_layout} and {!Bigarray.type-fortran_layout}
```
Same as above: `{!type:Bigarray.c_layout}` and `{!type:Bigarray.fortran_layout}`.

Octachron commented on stdlib/gc.mli:
```diff
@@ -394,7 +397,7 @@ val finalise : ('a -> unit) -> 'a -> unit
 
    The results of calling {!String.make}, {!Bytes.make}, {!Bytes.create},
-   {!Array.make}, and {!Stdlib.ref} are guaranteed to be
+   {!Array.make}, and {!Stdlib.val-ref} are guaranteed to be
```
`{!val:Stdlib.ref}`

Octachron commented on stdlib/lexing.mli:
```diff
@@ -187,7 +187,7 @@ val flush_input : lexbuf -> unit
 (**/**)
 
-(** {1  } *)
+(** {1 Internal functions} *)
```
I would propose to delete this section to be consistent with the changes in `parser.mli`.

Octachron commented on stdlib/stdlib.mli:
```diff
@@ -1284,7 +1284,7 @@ type ('a,'b) result = Ok of 'a | Error of 'b
       [out_channel];
       for [printf]-style functions from module {!Format}, ['b] is typically
-      {!Format.formatter};
+      {!Format.type-formatter};
```
`{!type:Format.formatter}`

Octachron commented on utils/diffing.mli:
```diff
@@ -14,7 +14,7 @@
 (**************************************************************************)
 
-(** {0 Parametric diffing}
+(** {1 Parametric diffing}
```
That header was intended to be the module title, isn't that the `0` level in odoc?

Julow replied:
Odoc doesn't allow to write a heading of level 0:
```
File "utils/diffing.mli", line 17, characters 4-26:
Warning: '{0': heading level should be lower than top heading level '0'.
```

Do you think it should allow that if it's at the top of a module (before any code and doc) ?
Odoc currently generate an equivalent of a level 0 heading, "Module Diffing".

Octachron replied:
Ah right, this is a difference between `mld` and `mli`, isn't it? It might be better to enforce some homogeneity.

In this case, the simplest correction is probably to transform this heading into a preamble rather than lower its level (or remove it altogether).

Julow replied:
`mld` requires a level-0 heading, so it's a bit inconsistent but I'm not sure it's a good idea to allow redefining the title for modules. I've opened an issue in Odoc just in case, https://github.com/ocaml/odoc/issues/815.

Octachron replied:
Ah sorry, I meant the reverse: that it was indeed better to enforce some homogeneity on titles for modules.

Octachron commented on utils/local_store.mli:
```diff
@@ -23,8 +23,8 @@
 val s_ref : 'a -> 'a ref
-(** Similar to {!ref}, except the allocated reference is registered into the
-    store. *)
+(** Similar to {!val-ref}, except the allocated reference is registered into
```
`{!val:ref}`

Octachron commented on utils/local_store.mli:
```diff
@@ -63,4 +63,4 @@ val reset : unit -> unit
 
 val is_bound : unit -> bool
 (** Returns [true] when a scope is active (i.e. when called from the callback
```
It looks like the active scope should be replaced by an active store here too.

Octachron commented on api_docgen/Makefile:
```diff
@@ -12,20 +12,33 @@
+install: ocamldoc-install
+texi: ocamldoc-texi
+pdf: ocamldoc-pdf
+endif
```
Since the main Makefile is not including `Makefile.common` anymore, you probably to define the `all` target explicitly here
```
all: html pdf man
```

Julow replied:
The targets are a bit different depending on the backend, I've added the `all` target but it's not at the top. Should it be?

Octachron replied:
The `all` target is more a test target, it should not be at the top.

Octachron commented:
Once my handful or remarks (and the merge conflict) are fixed, I think this PR will be ready to be merged.

Julow committed:
Use ocamldoc reference syntax

Co-authored-by: Florian Angeletti <florian.angeletti@inria.fr>

xavierleroy commented:
Will the updated documentation comments still work with the venerable ocamldoc?  Are we switching all documentation generation (to HTML, LaTeX and man pages) to odoc at the same time this PR is merged?  I'm a bit confused here.

Octachron commented:
The plan is that the updated comments will be fully valid for both ocamldoc and odoc.
We will still support both ocamldoc and odoc for documentation generation after this PR.

Before this PR, odoc would raise warnings on some documentation comments because odoc is stricter than ocamldoc in term of ambiguous references or in term of  syntax for both formatting and documentation tags.

The changes in this PR can summarized into 3 categories

* disambiguated references
* formatting tags start and stop on a new line
* documentation tags start on a new line
* more precise dependency tracking for documentation files
* some reorganization of the api_docgen's `Makefile`s (I imagine that it is hard to resist to the temptation of changing those)

Julow commented:
> Will the updated documentation comments still work with the venerable ocamldoc?

Yes, Odoc is still disabled at configure time and ocamldoc is still used to build the documentation.

> some reorganization of the api_docgen's Makefiles

I couldn't resist. The dependencies between the compile commands need to be more precise, so it's a bit justified.

Julow committed:
Tweaks to the documentation

Octachron reviewed (approved):
Improving odoc support while keeping compatibility seems the way to go for 4.14 .
